src/FileSystem/FileSystem.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace OomphInc\WASP\FileSystem;
+
+use RuntimeException;
+
+class FileSystem implements FileSystemInterface {
+
+	use DirStackTrait;
+
+	/**
+	 * @param string [$cwd] optional starting current working directory
+	 */
+	public function __construct($cwd = null) {
+		$this->setCwd($cwd);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function resolvePath($path) {
+		// default path if null provided
+		if ($path === null) {
+			return getcwd();
+		}
+
+		// is the path relative?
+		if (!preg_match('#^/|[a-z]+://#', $path)) {
+			$path = $this->getCwd() . '/' . $path;
+		}
+
+		return is_readable($path) ? realpath($path) : $path;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getFiles($pattern, $relative = true) {
+		$cwd = $this->getCwd();
+		$files = [];
+		foreach (glob($cwd . '/' . $pattern, GLOB_BRACE) as $file) {
+			// strip off the files root path
+			$files[] = $relative ? preg_replace('#^' . preg_quote(rtrim($cwd, '/') . '/', '#') . '#', '', $file) : $file;
+		}
+		return $files;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function fileExists($path) {
+		return is_readable($this->resolvePath($path));
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function readFile($path) {
+		$file = file_get_contents($this->resolvePath($path));
+		if ($file === false) {
+			throw new RuntimeException("Could not read file '$path'");
+		}
+		return $file;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function writeFile($path, $contents) {
+		if (file_put_contents($this->resolvePath($path), $contents) === false) {
+			throw new RuntimeException("Could not write to file '$path'");
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function deleteFile($path) {
+		if ($this->fileExists($path)) {
+			if (!unlink($this->resolvePath($path))) {
+				throw new RuntimeException("Could not delete file '$path'");
+			}
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function renameFile($oldPath, $newPath) {
+		if ($this->fileExists($oldPath)) {
+			if (!rename($this->resolvePath($oldPath), $this->resolvePath($newPath))) {
+				throw new RuntimeException("Could not rename file '$oldPath'");
+			}
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function mkDir($path) {
+		if (!mkdir($this->resolvePath($path), 0777, true)) {
+			throw new RuntimeException("Could not create directory '$path'");
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function rmDir($path) {
+		exec('rm -rf ' . escapeshellarg($this->resolvePath($path)), $output, $exitCode);
```

Comment:
This is probably outside of the scope for this PR, but if we see this existing as a functions file generator in addition to a part of a real build process, then I think we should do our best to support windows and implement this with PHP functions (and start using `DIRECTORY_SEPARATOR` instead of `'/'`).

Reply:
ugh, yeah, you're right - it's just that there isn't an easy way to remove a whole directory in PHP without iterating on a file structure and deleting each file first

Reply:
I didn't test this, but here ya go
```php
public function rm($path) {
	$path = $this->resolvePath($path);
	foreach (scandir($path) as $file) {
		if (is_dir($file)) {
			$this->rm($file);
			$success = rmdir($file);
		} else {
			$success = unlink($file);
		}
	}
	if (!$success) {
		throw new RuntimeException("Could not delete '$path'");
	}
}```

Reply:
you're making it real hard to be lazy